apple/internal/resource_rules/apple_metal_library.bzl

```
@@ -0,0 +1,173 @@
+    if platform_string == "macos":
+        platform_string = "macosx"
+
+    environment = "" if platform.is_device else "-simulator"
+
+    return "air64-apple-{platform}{version}{environment}".format(
+        environment = environment,
+        platform = platform_string,
+        version = target_os_version,
+    )
+
+def _apple_metal_library_impl(ctx):
+    air_files = []
+
+    platform_prerequisites = platform_support.platform_prerequisites(
+        apple_fragment = ctx.fragments.apple,
+        build_settings = ctx.attr._xplat_toolchain[AppleXPlatToolsToolchainInfo].build_settings,
+        config_vars = ctx.var,
+        device_families = None,
+        explicit_minimum_deployment_os = None,
+        explicit_minimum_os = None,
+        features = features_support.compute_enabled_features(
+            requested_features = ctx.features,
+            unsupported_features = ctx.disabled_features,
+        ),
+        objc_fragment = None,
+        platform_type_string = str(ctx.fragments.apple.single_arch_platform.platform_type),
+        uses_swift = False,
+        xcode_version_config = ctx.attr._xcode_config[apple_common.XcodeVersionConfig],
+    )
+    target = _metal_apple_target_triple(platform_prerequisites)
+
+    for src in ctx.files.srcs:
+        air_file = ctx.actions.declare_file(
+            paths.replace_extension(src.basename, ".air"),
+        )
+        air_files.append(air_file)
+
+        args = ctx.actions.args()
+        args.add("metal")
+        args.add("-c")
+        args.add("-target", target)
+        args.add("-o", air_file)
+        args.add_all(ctx.attr.copts)
+        args.add(src.path)
+
+        apple_support.run(
+            actions = ctx.actions,
+            apple_fragment = platform_prerequisites.apple_fragment,
+            arguments = [args],
+            executable = "/usr/bin/xcrun",
+            inputs = [src] + ctx.files.hdrs,
+            mnemonic = "MetalCompile",
+            outputs = [air_file],
+            xcode_config = platform_prerequisites.xcode_version_config,
+        )
+
+    out = ctx.actions.declare_file(ctx.attr.out)
+
+    args = ctx.actions.args()
+    args.add("metallib")
+    args.add("-o", out.path)
+    args.add_all(air_files)
+
+    apple_support.run(
+        actions = ctx.actions,
+        apple_fragment = platform_prerequisites.apple_fragment,
+        arguments = [args],
+        executable = "/usr/bin/xcrun",
+        inputs = air_files,
+        mnemonic = "MetallibCompile",
+        outputs = [out],
+        xcode_config = platform_prerequisites.xcode_version_config,
+    )
+
+    return [
+        DefaultInfo(files = depset([out])),
+        resources.bucketize_typed([out], "unprocessed"),
+    ]
+
+apple_metal_library = rule(
+    attrs = dicts.add(
+        apple_support.action_required_attrs(),
+        apple_toolchain_utils.shared_attrs(),
+        {
+            "copts": attr.string_list(
+                doc = """\
+A list of compiler options.
```

Could we add some more details to these, maybe:

```suggestion
A list of compiler options passed to the `metal` compiler for each source.
```